specification/dartLangSpec.tex

```diff
@@ -18117,35 +18118,76 @@ \subsection{Yield-Each}
   It is a dynamic type error if the class of $o$
   is not a subtype of \code{Stream<$T_f$>}.
 Otherwise
+\item
+  If the stream associated with $m$ has been cancelled,
+  then execution of $s$ completes by returning without a value.
+  \commentary{In this case, the \code{\YIELD*} operation does
+  not begin to listen to the stream $o$.}
+Otherwise
+\item
+  The $o$ stream is listened to by invoking \code{$v_o$.listen}
+  with system provided arguments,
+  where $v_o$ is a fresh variable referencing the stream $o$.
+  Let $u$ be the stream subscription returned by the call to
+  \code{$p$.listen} and let $v$ be a fresh variable bound to $u$.
+\item
+  If the stream associated with $m$ is paused, then $u$ is immediately
+  paused as if by invoking \code{$v$.pause} with no arguments.
+\item{}
+  If the stream associated with $m$ has been cancelled, then
+  $u$ is cancelled by executing \code{await $v$.cancel();}.
+  If this execution does not throw, the execution of $s$
+  completes by returning without a value.
+  % This can only happen if a call to "listen" or "pause" above has
+  % side effects which cancels the stream of $m$.
 \item
   The nearest enclosing asynchronous for loop (\ref{asynchronousFor-in}),
   if any, is paused.
 \item
-  The $o$ stream is listened to, creating a subscription $s$,
-  and for each event $x$, or error $e$ with stack trace $t$, of $s$:
+  Execution of $m$ is suspended.
+  While $m$ is suspended, the following reactions happen if $u$ delivers
+  an event or if the stream associated with $m$ is paused, resumed or cancelled.
+  \commentary{It is unspecified what happens if $u$ delivers events
+  before $m$ is suspended or after $m$ is resumed, or if $u$ delivers events
+  after a call to \code{$v$.cancel} or a call to \code{$v$.pause} which has
+  not been followed by a call to \code{$v$.resume}, or after having delivered
+  a done event.}
+  % Such events should be ignored.
   \begin{itemize}
   \item
-    If the stream $u$ associated with $m$ has been paused,
-    then execution of $m$ is suspended until $u$ is resumed or canceled.
+    If the stream associated with $m$ becomes paused,
+    then $u$ is paused by executing \code{$v$.pause();}.
+    It is unspecified what happens if this execution throws.
+    % The error can be passed to \code{Zone.current.handleUncaughtError}.
   \item
-    If the stream $u$ associated with $m$ has been canceled,
-    then $s$ is canceled by evaluating \code{\AWAIT{} v.cancel()}
-    where $v$ is a fresh variable referencing the stream subscription $s$.
-    Then, if the cancel completed normally,
-    the stream execution of $s$ returns without an object
-    (\ref{statementCompletion}).
+    If the stream associated with $m$ is resumed after being paused,
+    then $u$ is resumed as by executing \code{$v$.resume();}.
```

Comment:
Could again be `evaluating \code{$v$.resume()}.`

Reply:
ACK